pkg/cmd/cli/cmd/rollback.go

```diff
@@ -38,151 +43,267 @@ a human-readable representation of the updated deployment configuration instead
+	for _, trigger := range rolledback.Triggers {
+		disabled := []string{}
+		if trigger.Type == deployapi.DeploymentTriggerOnImageChange && !trigger.ImageChangeParams.Automatic {
+			disabled = append(disabled, trigger.ImageChangeParams.From.Name)
+		}
+		if len(disabled) > 0 {
+			reenable := fmt.Sprintf("oc deploy %s --enable-triggers", rolledback.Name)
+			fmt.Fprintf(o.out, "Warning: the following images triggers were disabled: %s\n  You can re-enable them with: %s\n", strings.Join(disabled, ","), reenable)
+		}
+	}
+
 	return nil
 }
 
-// Update persists the given DeploymentConfig.
-func (r *helper) Update(config *deployapi.DeploymentConfig) (*deployapi.DeploymentConfig, error) {
-	return r.updateConfig(config.Namespace, config)
+// findResource tries to find a deployment or deploymentconfig named
+// targetName using a resource.Builder. For compatibility, if the resource
+// name is unprefixed, treat it as an rc first and a dc second.
+func (o *RollbackOptions) findResource(targetName string) (runtime.Object, error) {
+	candidates := []string{}
+	if strings.Index(targetName, "/") == -1 {
+		candidates = append(candidates, "rc/"+targetName)
+		candidates = append(candidates, "dc/"+targetName)
+	} else {
+		candidates = append(candidates, targetName)
+	}
+	var obj runtime.Object
+	for _, name := range candidates {
+		r := o.getBuilder().
+			NamespaceParam(o.Namespace).
+			ResourceTypeOrNameArgs(false, name).
+			SingleResourceType().
+			Do()
+		if r.Err() != nil {
+			return nil, r.Err()
+		}
+		resultObj, err := r.Object()
+		if err != nil {
+			// If the resource wasn't found, try another candidate.
+			if kerrors.IsNotFound(err) {
+				continue
+			}
+			return nil, err
+		}
+		obj = resultObj
+		break
+	}
+	if obj == nil {
+		return nil, fmt.Errorf("%s is not a valid deployment or deploymentconfig", targetName)
+	}
+	return obj, nil
+}
+
+// findTargetDeployment finds the deployment which is the rollback target by
+// searching for deployments associated with config. If desiredVersion is >0,
+// the deployment matching desiredVersion will be returned. If desiredVersion
+// is <=0, the last completed deployment which is older than the config's
+// version will be returned.
+func (o *RollbackOptions) findTargetDeployment(config *deployapi.DeploymentConfig, desiredVersion int) (*kapi.ReplicationController, error) {
+	// Find deployments for the config sorted by version descending.
+	deployments, err := o.kc.ReplicationControllers(config.Namespace).List(deployutil.ConfigSelector(config.Name))
+	if err != nil {
+		return nil, err
+	}
+	sort.Sort(deployutil.DeploymentsByLatestVersionDesc(deployments.Items))
+
+	// Find the target deployment for rollback. If a version was specified,
+	// use the version for a search. Otherwise, use the last completed
+	// deployment.
+	var target *kapi.ReplicationController
+	for _, deployment := range deployments.Items {
+		version := deployutil.DeploymentVersionFor(&deployment)
```

Comment:
As long as the labels are properly namespaced, we're set.  If they're not,
we need the plan to namespace them (which I assume was one of the spawned
issues).



On Jul 10, 2015, at 10:13 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rollback.go
<https://github.com/openshift/origin/pull/3643#discussion_r34357353>:

> +// is <=0, the last completed deployment which is older than the config's
> +// version will be returned.
> +func (o *RollbackOptions) findTargetDeployment(config *deployapi.DeploymentConfig, desiredVersion int) (*kapi.ReplicationController, error) {
> +	// Find deployments for the config sorted by version descending.
> +	deployments, err := o.kc.ReplicationControllers(config.Namespace).List(deployutil.ConfigSelector(config.Name))
> +	if err != nil {
> +		return nil, err
> +	}
> +	sort.Sort(deployutil.DeploymentsByLatestVersionDesc(deployments.Items))
> +
> +	// Find the target deployment for rollback. If a version was specified,
> +	// use the version for a search. Otherwise, use the last completed
> +	// deployment.
> +	var target *kapi.ReplicationController
> +	for _, deployment := range deployments.Items {
> +		version := deployutil.DeploymentVersionFor(&deployment)

 If you're concerned deployutil.ConfigSelector can return non-deployment
RCs, we'll need to add more abstraction which combines the selection with
further validations/filtering and plumb that throughout the deployment code.

If that's what it will take, then lets open an issue to get that done. I
got burned too badly by RC->Pod label selecting matching to consider this a
reasonable and innocuous risk.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3643/files#r34357353>.